### PR TITLE
passing data between vcs using delegate and protocol

### PR DESCRIPTION
**Pull request by jimmygtz15**

### CODE REVIEW

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  JimmySecondViewController.swift
+//  BackToBasicsUIKit
+//
+//  Created by Jimmy Gutierrez on 5/23/19.
+//  Copyright © 2019 SwiftMiami. All rights reserved.
+//
+
+import UIKit
+
+protocol canRecieve {
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:11]: Names of protocols, classes, structs, etc should all be capitalised. The name should also be descriptive, and spelled correctly (`receive`).

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        label.text = data
+
+    }
+    
+    
+    @IBAction func passItBack(_ sender: Any) {
+       delegate?.dataRecieved(data: textField.text!)
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:32]: You should guard that the text field actually contains text, rather than force unwrapping.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        self.navigationController?.popViewController(animated: true)
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:33]: `self` isn't necessary.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    func dataRecieved(data: String)
+}
+
+class JimmySecondViewController: UIViewController {
+    
+    var delegate : canRecieve?
+    
+    var data = ""
+    
+    @IBOutlet var label: UILabel!
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:21]: IBOutlets should be declared as `private` (we don't want outsiders to modify them) and `weak`.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        super.viewDidLoad()
+        label.text = data
+
+    }
+    
+    
+    @IBAction func passItBack(_ sender: Any) {
+       delegate?.dataRecieved(data: textField.text!)
+        self.navigationController?.popViewController(animated: true)
+//        dismiss(animated: true, completion: nil)
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:34]: Code that isn't necessary should be removed to avoid confusion.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+//
+//  Created by Jimmy Gutierrez on 5/23/19.
+//  Copyright © 2019 SwiftMiami. All rights reserved.
+//
+
+import UIKit
+
+class JimmyViewController: UIViewController, canRecieve {
+
+    @IBOutlet var label: UILabel!
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:13]: IBOutlets should be declared as private (we don't want outsiders to modify them) and weak.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @IBOutlet var textField: UITextField!
+    
+    override func viewDidLoad() {
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:16]: If we're not overriding, we should remove the functionalists.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+
+    }
+    
+    
+    @IBAction func buttonPressed(_ sender: Any) {
+        performSegue(withIdentifier: "sendDataForwards", sender: self)
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:23]: Instead of using a string literal for the identifier here and in the method below, it might be better to declare a private variable to avoid spelling mistakes and the ability to rename this identifier in a single place.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "sendDataForwards" {
+            
+            let secondVC = segue.destination as! JimmySecondViewController
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:29]: A `guard` > force unwrapping.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            
+            secondVC.data = textField.text!
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:31]: The guard above should also check for the text field contents, rather than force unwrapping.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            
+            secondVC.delegate = self
+        }
+    }
+    
+    func dataRecieved(data: String) {
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:37]: This is totally fine, though I am a fan of creating a separate extension that conforms to the delegate protocol and have the delegate methods there for tidiness.

File: BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift [hunk 12]
```diff
@@ -66,11 +67,19 @@ class MainTableViewController: UITableViewController {
...
+        case 10: demoVC = JimmyViewController()
+            
         default:
             assertionFailure("Cell not configured yet")
         }
-
+        
+        if indexPath.row == 10 {
+            let storyboard = UIStoryboard(name: "Jimmy", bundle: nil)
+            let destination = storyboard.instantiateViewController(withIdentifier: "JimmyViewController") as! JimmyViewController
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift:78]: 💂

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                                <rect key="frame" x="20" y="314" width="374" height="40"/>
+                                <constraints>
+                                    <constraint firstAttribute="height" constant="40" id="9zj-IF-OEb"/>
+                                </constraints>
+                                <nil key="textColor"/>
+                                <fontDescription key="fontDescription" type="system" pointSize="25"/>
+                                <textInputTraits key="textInputTraits"/>
+                            </textField>
+                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="Label" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="ASx-aW-Sso">
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard:45]: You may want to add leading/trailing constraints to avoid the text from disappearing off screen if the label contents are very long.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+protocol canRecieve {
+    func dataRecieved(data: String)
+}
+
+class JimmySecondViewController: UIViewController {
+    
+    var delegate : canRecieve?
+    
+    var data = ""
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:19]: Renaming to `initialData` may make it a little clearer what the role of this variable is.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+//  BackToBasicsUIKit
+//
+//  Created by Jimmy Gutierrez on 5/23/19.
+//  Copyright © 2019 SwiftMiami. All rights reserved.
+//
+
+import UIKit
+
+protocol canRecieve {
+    func dataRecieved(data: String)
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:12]: The function signature should also include the sender, so the delegate knows who called it:

`func dataRecieved(from viewController: UIViewController, data: String)`

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    @IBOutlet var textField: UITextField!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        label.text = data
+
+    }
+    
+    
+    @IBAction func passItBack(_ sender: Any) {
```
Comment (paulofierro) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:31]: While this pattern works, it assumes that the previous view controller is the same as its delegate. A slightly better pattern may be to send the data back to the delegate and then allow the delegate to control any navigation (in this case popping the VC).

However, a _much better_ approach is to use unwind segues here as they handle all of this for you, without the need of a delegate protocol or handling the navigation yourself. This is a great tutorial on this type of segue: https://matteomanferdini.com/unwind-segue/

**Review by paulofierro — CHANGES_REQUESTED** [review]

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 3 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:33]: Agreed. You only need to use `self` when dealing with a closure (for example in a closure after a WebService response comes in.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 4 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:21]: Private makes sense. @ivancr regarding this, should IBOutlets be weak or strong, and why?

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 5 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:34]: @jimmygtz15 Instead of commenting, can just delete the commented code.

File: BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift (hunk 12 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift:78]: Instead of the if statement, you can attach the if statement action to the switch statement and then return, example:

`case 10: 
let storyboard = UIStoryboard...


return`

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard (hunk 13 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard:45]: Give it some constraints

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 14 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:19]: Always important for names to make sense.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 16 above)
Comment (gcharles-chwy) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:31]: Didn't know that! Awesome!

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 4 above)
Comment (paulofierro) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:21]: Technically VC is not the owner of the label, the VC's view is and the VC already has a strong reference to its view.

If the VC also kept strong references to the view's children (e.g. this label) then those won't be deallocated if the view was to get swapped out. There may be use cases where you'd want a `strong` outlet, but I haven't run into one yet. YMMV 😃

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard (hunk 13 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/Jimmy.storyboard:45]: Or better yet, get rid of the storyboard and do it programmatically. You can pair with @gionoa and he'll show you the ways of the force.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 1 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:11]: I would add that even then I find this name wrong. The typical naming convention for something like this is `{name of the VC}Delegate`. In this case it would be: `protocol JimmySecondViewControllerDelegate: class {`

While delegates use protocols they have a specific use case and conventions. When you work in teams following platform and community conventions is super important.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 15 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:12]: I know that you (Jimmy) were trying to follow the protocol pattern of a capability but I think you should scrap this idea and think of it as a Delegate specific to `JimmySecondViewController`. Then the function should be a specific need of `JimmySecondViewController` to communicate to another object. Maybe `func didEnterText(_ text: String)`.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+import UIKit
+
+protocol canRecieve {
+    func dataRecieved(data: String)
+}
+
+class JimmySecondViewController: UIViewController {
+    
+    var delegate : canRecieve?
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:17]: Let me introduce you to memory leaks. Let me know if you see why you have memory leak here and how you'd fix it.

Also the space should only go after the `:`. Like this: 
```suggestion
    var delegate: canRecieve?
```

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 14 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:19]: I would also go further and add that using the noun data makes this too generic. This is text we're talking here, it's a very specific type of data. The name should reflect not only in its use case but also its identity.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 4 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:21]: Apple recommends that in the ARC age your IBOutlets should be strong:
https://stackoverflow.com/questions/7678469/should-iboutlets-be-strong-or-weak-under-arc

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    @IBOutlet var label: UILabel!
+    @IBOutlet var textField: UITextField!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        label.text = data
+
+    }
+    
+
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:30]: Please remove all these empty lines. It may sound OCD but being neat with your code is very important. You're sharing this codebase with a team, please clean up after yourself.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 16 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:31]: I'm sure there's somebody out there using these and they have their merits but I've never seen or work in any team that used them. I'd highly recommend you go the Delegate pattern.

Also, each part of the whole system should be named accordingly. This IBAction maybe happens when you tap a button, then make the name say that. What you do inside is another story for me.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 2 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:32]: Indeed, there's very few instances where it does make sense to force unwrap this is not one of them.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    @IBAction func passItBack(_ sender: Any) {
+       delegate?.dataRecieved(data: textField.text!)
+        self.navigationController?.popViewController(animated: true)
+//        dismiss(animated: true, completion: nil)
+        
+    }
+    
+
+    
+}
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:40]: Please delete empty lines too.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+    @IBOutlet var label: UILabel!
+    @IBOutlet var textField: UITextField!
+    
+    override func viewDidLoad() {
+
+
+    }
+    
+
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:21]: Empty spaces

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift (hunk 8 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:23]: Or just don't use segues, they're gross.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift (hunk 10 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:31]: If `JimmySecondViewController` requires a string to be used, then it should initialized with it.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift (hunk 11 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmyViewController.swift:37]: Agreed, this is pretty much a community standard. It should go at the end of the VC something like this:
```
extension JimmyViewController: JimmySecondViewController {

}
```

File: BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift (hunk 12 above)
Comment (ivancr) [reply BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift:78]: Don't use a storyboard. Another important lesson to learn here is that you're working on a codebase with a team and you should conform to the style and architecture of it. Nobody else is using storyboards and you shouldn't either.
Be a nice team citizen and conform to the style or convince the team that we should change the style. Don't go rogue though.

**Review by ivancr — CHANGES_REQUESTED** [review]
You're doing great man, please don't be discouraged by any comments. There's no progress if there's pain, embrace the growing pains.

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift (hunk 4 above)
Comment (paulofierro) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Jimmy/JimmySecondViewController.swift:21]: @ivancr i watched that session and was curious as to why they said that, as their documentation, Xcode defaults and sample code reflects the opposite.

I like this roundup: https://cocoacasts.com/should-outlets-be-weak-or-strong/

TLDR: It doesn't _really_ matter — just be consistent.

**Comment by jimmygtz15** [comment]
Thank you for taking the time to review this and providing your input.🙏
I'll send my update when I'm ready.
